src/peergos/server/net/DHTHandler.java

```diff
@@ -73,7 +81,50 @@ public void handle(HttpExchange httpExchange) throws IOException {
                     List<byte[]> data = MultipartReceiver.extractFiles(httpExchange.getRequestBody(), boundary);
                     boolean isRaw = last.apply("format").equals("raw");
 
-                    (isRaw ? dht.putRaw(writer, data) : dht.put(writer, data)).thenAccept(hashes -> {
+                    // check writer is allowed to write to this server, and check their free space
+                    if (! keyFilter.isAllowed(writerHash))
+                        throw new IllegalStateException("Key not allowed to write to this server: " + writerHash);
+
+                    // Get the actual key, unless this is the initial write of the signing key during sign up
+                    // In the initial put of a signing key during sign up the key signs itself (we still check the hash
+                    // against the core node)
+                    Supplier<PublicSigningKey> fromDht = () -> {
+                        try {
+                            return PublicSigningKey.fromCbor(dht.get(writerHash.hash).get().get());
+                        } catch (Exception e) {
+                            throw new RuntimeException(e);
+                        }
+                    };
+                    Supplier<PublicSigningKey> inBandOrDht = () -> {
+                        try {
+                            // Check if this is possible to cause an OOM error before parsing it as cbor
+                            if (PublicSigningKey.maybeValidKey(data.get(0))) {
+                                PublicSigningKey candidateKey = PublicSigningKey.fromByteArray(data.get(0));
+                                PublicKeyHash calculatedHash = dht.hashKey(candidateKey);
+                                if (calculatedHash.equals(writerHash)) {
+                                    candidateKey.unsignMessage(ArrayOps.concat(signatures.get(0), data.get(0)));
+                                    return candidateKey;
+                                }
+                            }
+                        } catch (Throwable e) {
```

Comment:
Catch something more specific?

Reply:
I tried that and it resulted in 5 different exception types before I gave up. In this case I don't care what went wrong, just that something went wrong.